### ISSUE TITLE
how to understand the  output of discriminator  

### ISSUE THREAD

**Issue by hiterjoshua** [jantic/DeOldify#263]
@jantic Hi there! I want to see the output of the discriminator, and it really makes me lost, can you give me some tips? 
the picture input has a size (400,400,3) and I get a list of image [1,484] as result, the length of the list is 3. As far as I know, discriminator would usually have a float number between 0 and 1......

![image](https://user-images.githubusercontent.com/29330657/89992307-abf04e80-dcb7-11ea-97ef-79fd29623481.png)

**Comment by jantic** [comment]
I guess the easiest way to think of it is that the burden of turning this into a single score shifts to the critic loss function itself- [BCEWithLogitsLoss](https://pytorch.org/docs/master/generated/torch.nn.BCEWithLogitsLoss.html). If you look at the source code in FastAI AdaptiveLoss wraps this loss to make sure that the loss's targets are fed with the appropriate length tensor filled with zeros for the "fake" predictions and ones for the "real" predictions.  It's a bit hard to follow but the code for that is [here](https://github.com/fastai/fastai/blob/master/fastai/vision/gan.py) and perhaps starting at gan_loss_from_func would help in understanding.  

Now as to why they decided to do it this way in the FastAI code?  I actually never asked! I would suspect it's more numerically robust and stable (that has been my assumption), but I never actually tested that theory because I moved away from GANs in my DeOldify work soon after I completed the initial rollout of the current open source version here.   But basically you can think of it as a round about way of achieving the same exact thing that you're looking for here.

**Comment by hiterjoshua** [comment]
@jantic Thank you for your reply, now I understand how to transform the list to a float between 0 and 1 by BCEWithLogitsLoss , and make it perform as a loss function with  gan_loss_from_func.

but why the output is a list of which  the length is 3? According to the discriminator model, the pic input of size  (400,400,3) would  have a flatten output of (1,484), while I receive a output of [(1,484), (1,484), (1,484)].

My code is below:
The model part:
```
def colorize_crit_learner(
    data: ImageDataBunch,
    loss_critic=AdaptiveLoss(nn.BCEWithLogitsLoss()),
    nf: int = 256,
) -> Learner:
    return Learner(
        data,
        custom_gan_critic(nf=nf),
        metrics=accuracy_thresh_expand,
        loss_func=loss_critic,
        wd=1e-3,
    )

# Weights are implicitly read from ./models/ folder
def critic_inference_wide(
    root_folder: Path, weights_name: str, nf_factor: int = 256) -> Learner:
    data = get_dummy_databunch()
    learn = colorize_crit_learner(data=data, loss_critic=AdaptiveLoss(nn.BCEWithLogitsLoss()), nf=256)
    learn.path = root_folder
    learn.model_dir = os.path.dirname(os.path.abspath(__file__))
    learn.load(weights_name)
    #print(learn)
    learn.model.eval()
    return learn
```

The inference part:

        model_image= PIL.Image.fromarray(np.zeros((400,400),np.uint8)).convert('RGB')
        print(numpy.asarray(model_image))
        
        x = pil2tensor(model_image, np.float32)
        x = x.to(self.device)
        x.div_(255)
        x, y = self.norm((x, x), do_x=True)

        try:
            result = self.learn.pred_batch(
                ds_type=DatasetType.Valid, batch=(x[None], y[None]), reconstruct=True
            )
        except RuntimeError as rerr:
            if 'memory' not in str(rerr):
                raise rerr
            print('Warning: render_factor was set too high, and out of memory error resulted. Returning original image.')
            return model_image
        
        print(len(result), type(result), result)`

The two line print output is:
`(400, 400, 3)`

`3 <class 'list'> [Image (1, 484), Image (1, 484), Image (1, 484)]`

Can you give me some tips why the output of discriminator is three (1,484) instead of one (1,484)? Hope your answer, Thank you!
